Feature: Custom Alphabet Option

GianniBYoung opened the pull request:
This PR introduces the option to add a custom alphabet as requested and resolves #3. 

The option can be set in the following way `require "quick_buffer_jump".setup{ alphabet = "qwerty" }` and will default to `"abcdefghijklmnopqrstuvwxyz"` if not specified.

Commit:
Configuration option for alphabet

Commit:
simplifying options

Commit:
updated README to include example for setting options

sathishmanohar commented on lua/quick_buffer_jump.lua:
```diff
@@ -1,5 +1,13 @@
 local M = {}
 
+M.config = {
+    alphabet = "abcdefghijklmnopqrstuvwxyz",
```
Thanks for the PR!

Please note `jkq` are reserved alphabets. They are essential for internal functionality of the plugin.

These alphabets are safe to use `abcdefghilmnoprstuvwxyz`

GianniBYoung replied:
I did not realize. I have updated the default alphabet option to be correct and included code to check for the existence of the reserved characters and remove them if present!

Commit:
alphabet validation

The alphabet now checks for the characters j,k,q and will remove them from the user provided map if present

true return for alphabet validation

Commit:
updated README

sathishmanohar commented:
I tested the PR it works great. However there is a problem, not with the PR but with the plugin. When we are adding custom alphabets like `asdf` and when there are 8 open buffers. The alphabets are repeated. This breaks the plugin to those who frequently work with a good number of open buffers.

I think an implementation like, two premade configs to choose from and one custom alphabet config where users can set their own alphabet sequence would be ideal. Like below.

alphabetic = `abcdefghilmnoprstuvwxyz` (default)
ergonomic = `asdfghlentriowmucvybpxz` (chosen with a config flag but can't be customized)
custom = (user chosen letters that requires a minimum of 9 letters. reasoning given below)

(You can change the ergonomic letter ordering as you like.)

If the feature introduces a config to choose between pre-made alphabetic and ergonomic ordering approach. I can merge the PR immediately after testing.

I'm open to custom alphabets as well but I have to change the plugin behavior to loop through alphabets and show double alphabet combinations when the list is exhausted.

Even that approach has it's problems because when there are only 4 letters given. All two letter combinations will have a total of 16 combinations. So there has to be a bottom limit of 9 custom alphabets required so that 81 open buffers can work without any problems.

The two letter approach when implemented with the 23 alphabets we have now, It can support 23^2 that is 529 open buffers.

For now, I'd consider custom alphabets feature blocked because the plugin doesn't support two letter combinations. But I think just providing a pre-made ergonomic option to choose from will address this issue for most usecases.

Regardless of the technical limitations, I firmly believe in an option in which users don't  have to do a lot thinking and figuring out an ergonomic combinations. So pre-made ergonomic options is for those users and the custom alphabets can be implemented down the line for users who are very particular about their setup.

I sincerely welcome your changes. I hope you consider my feedback. Thanks.

GianniBYoung commented:
Sure thing, I will work to make the change to provide the alphabet ordering and the ergonomic ordering. 

As for addressing the current technical challenges I will leave that to another issue/future work.

 A different approach could be limiting the amount of buffers that are labeled for jumping and instead allowing the user to navigate to the target buffer using alternative techniques for extreme cases with a large amount of buffers.

Commit:
adding preconfigured `ergonomic_alphabet` option

Commit:
shift `g` over in alphabet order

Commit:
updated README

typo

GianniBYoung commented:
The changes have been added

sathishmanohar commented:
```
Error detected while processing /home/sathish/satdot/vim/.config/nvim/init.vim:
line  196:
E5108: Error executing lua ....vim/plugged/quick-buffer-jump/lua/quick_buffer_jump.lua:28: ergonomic_alphabet: expected boolean, got nil
stack traceback:
        [C]: in function 'error'
        vim/shared.lua: in function 'validate'
        ....vim/plugged/quick-buffer-jump/lua/quick_buffer_jump.lua:28: in function 'setup'
        [string ":lua"]:8: in main chunk
Press ENTER or type command to continue
```

When I try the plugin with `require "quick_buffer_jump".setup{ ergonmic_alphabet = true }` I get the above error. Is validation necessary at this stage? Since we are just choosing from one of two pre-made alphabet ordering at this point.

sathishmanohar commented:
> A different approach could be limiting the amount of buffers that are labeled for jumping and instead allowing the user to navigate to the target buffer using alternative techniques for extreme cases with a large amount of buffers.

Since users have no control over which buffers are labelled because our list is based on opened buffers. Labeling them based on buffers they opened first might not result in a good UX in my humble opinion. I think allowing for 23 buffers must be sufficient for most users. We can look into this limitation when it becomes a problem or when we have to deal with it with the user configurable custom alphabet feature.

Commit:
removing validation

Commit:
Allowing ergonmic_alphabet option to be nil

GianniBYoung commented:
I removed the larger validation function as it is not really need right now.

I left in the check that asserts `ergonmoic_alphabet` must be a boolean but made it so it also accepts `nil` that way a call like `require quick_buffer_jump.setup{}` is valid and will use the default alphabet.

sathishmanohar commented:
Great work! Thanks a lot for your contribution.

GianniBYoung commented:
No problem, thanks for making the plugin!
